### PR TITLE
move crd api into a separate go module

### PR DESCRIPTION
**Pull request by leelavg**
this helps for code that is depending on client-op CRDs to only pull the submodule rather than whole client-op and also remove controller-runtime from this new submodule.

### CODE REVIEW

**Comment by leelavg** [comment]
/hold

can only review first commit, compilation is working, doing a test for runtime type registration atm.

**Comment by leelavg** [comment]
/unhold

tested `make run`
```
2024-09-17T05:29:00Z    INFO    Starting Controller     {"controller": "configmap", "controllerGroup": "", "controllerKind": "ConfigMap"}
2024-09-17T05:29:01Z    INFO    Starting workers        {"controller": "storageclaim", "controllerGroup": "ocs.openshift.io", "controllerKind": "StorageClaim", "worker count": 1}
2024-09-17T05:29:01Z    INFO    Starting workers        {"controller": "storageclient", "controllerGroup": "ocs.openshift.io", "controllerKind": "StorageClient", "worker count": 1}
2024-09-17T05:29:01Z    INFO    Starting workers        {"controller": "configmap", "controllerGroup": "", "controllerKind": "ConfigMap", "worker count": 1}
```

File: Makefile [hunk 1]
```diff
@@ -56,7 +56,10 @@ lint: ## Run golangci-lint against code.
 	$(IMAGE_BUILD_CMD) run --rm -v $(PROJECT_DIR):/app -w /app $(GO_LINT_IMG) golangci-lint run ./...
 
 godeps-update:  ## Run go mod tidy & vendor
+	@echo "Running godeps-update"
 	go mod tidy && go mod vendor
+	@echo "Running godeps-update on api submodule"
+	cd api && go mod tidy && go mod vendor
```
Comment (rewantsoni) [inline Makefile:62]: Could you update godeps-verify script as well [here](https://github.com/red-hat-storage/ocs-client-operator/blob/main/hack/godeps-verify.sh#L3)

File: go.mod [hunk 2]
```diff
@@ -2,10 +2,9 @@ module github.com/red-hat-storage/ocs-client-operator
 
 go 1.22.0
 
-toolchain go1.22.3
-
 replace (
 	github.com/portworx/sched-ops => github.com/portworx/sched-ops v0.20.4-openstorage-rc3 // required by Rook v1.12
+	github.com/red-hat-storage/ocs-client-operator/api => ./api
```
Comment (rewantsoni) [inline go.mod:7]: Should we move this to a separate replace block?

**Review by rewantsoni — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (leelavg) [reply Makefile:62]: done.

File: go.mod (hunk 2 above)
Comment (leelavg) [reply go.mod:7]: doesn't make any diff, nack.

File: api/go.mod [hunk 3]
```diff
@@ -0,0 +1,22 @@
+module github.com/red-hat-storage/ocs-client-operator/api
+
+go 1.22.0
```
Comment (Madhu-1) [inline api/go.mod:3]: Update this to 1.22.5 as well

File: api/v1alpha1/groupversion_info.go [hunk 4]
```diff
@@ -20,17 +20,48 @@ limitations under the License.
...
+// copied from https://github.com/kubernetes-sigs/controller-runtime/blob/2eb879/pkg/scheme/scheme.go
+// Copyright 2018 The Kubernetes Authors.
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// gist, controller-runtime scheme allows for gradual registration of types vs kube runtime which is more suited
+// for registering all types during construction of the builder in addition to providing gradual registration which is
+// used by controller-runtime and required parts are copied as below.
+
+// +kubebuilder:object:generate=false
+// Builder builds a new Scheme for mapping go types to Kubernetes GroupVersionKinds.
```
Comment (nb-ohad) [inline api/v1alpha1/groupversion_info.go:48]: 1. We should not have copywrite comment in the middle of a file.
2. I don't really understand the need for this change

File: api/go.mod (hunk 3 above)
Comment (nb-ohad) [reply api/go.mod:3]: Let's have this update as part of its own commit. The implications on the codebase might be large

Comment (leelavg) [reply api/go.mod:3]: Updating version here is bumping api-machinery which inturn bumping client-op deps so not to have more than what's required I didn't change version here.

**Review by nb-ohad — CHANGES_REQUESTED** [review]

**Comment by leelavg** [comment]
/hold

for two reviews.

File: api/v1alpha1/groupversion_info.go (hunk 4 above)
Comment (leelavg) [reply api/v1alpha1/groupversion_info.go:48]: 1. the code is copyrighted by Kubernetes and licensed w/ apache and so just preserving that note here, I could move this to a separate file though to be clean
2. either this or something similar to https://github.com/ceph/ceph-csi-operator/pull/105 be done, I chose the former as not to take away the flexibility of registering API types in their individual files

**Pull request merged**